docs/design.md

```diff
@@ -38,13 +58,164 @@ spec:
+
+Non-root users should be enforced using the `runAsNonRoot: true` field within Pod security contexts.
+
+### Storage
+
+#### Buckets
+
+In Kubernetes the combination of the following fields make up a unique reference to a given object (i.e. database key): `split(.apiVersion, "/")[0] + .kind + .metadata.namespace + .metadata.name`. By storing related artifacts using a similar scheme, restore operations are made trivial: the administrator will not need to worry about backing up the `.status.url` field (as opposed to a scheme where storage location is `.metadata.uid` based - which would be unique for a given in-time instance of an object). When deleting and recreating clusters, if the storage bucket persisted, objects (i.e. Models, Datasets) can simply be re-applied into the new cluster (i.e. [velero](https://velero.io/)/similar is not needed). This plays nicely with GitOps.
+
+Process:
+
+Calculate a hash.
+
+Hashes are used to [prevent bucket hot-spots](https://cloud.google.com/blog/products/gcp/optimizing-your-cloud-storage-performance-google-cloud-performance-atlas).
+
+As opposed to a random id, a hash is used to provide a deterministic pattern storing and finding artifacts. By including a cluster name, Substratus can optionally store artifacts across multiple clusters in a single bucket if a compelling use-case arises in the future.
+
+```sh
+hash_input = "clusters/{cluster}/namespaces/{namespace}/{resource_plural}/{name}"
+
+hash = sha256(hash_input)
+```
+
+For example, with the following Model, with a default Substratus installation (cluster = `substratus-1`):
+
+```yaml
+kind: Model
+metadata:
+  name: falcon-7b
+  namespace: team-a
+```
+
+The hash input string would be: `clusters/substratus-1/namespaces/team-a/models/falcon-7b`.
+
+The resulting MD5 hash would be: `f94a0d128bcbd9c1b824e9e5572baf86`.
+
+The following scheme can be used for storing artifacts for Models, Datasets, and Notebooks:
+
+```sh
+# Models
+gs://{bucket}/{hash}/model # Model artifacts (*.pt, etc)
+gs://{bucket}/{hash}/logs  # Logs and converted notebooks
+
+# Datasets
+gs://{bucket}/{hash}/data # Model artifacts (*.pt, etc)
+gs://{bucket}/{hash}/logs # Logs and converted notebooks
+
+# Notebooks
+gs://{bucket}/{hash}/build/{md5-checksum}.tar # Uploaded build context
+```
+
+The example Model's backing storage would end up being:
+
+```sh
+gs://abc123-substratus-artifacts/f94a0d128bcbd9c1b824e9e5572baf86/model/
+gs://abc123-substratus-artifacts/f94a0d128bcbd9c1b824e9e5572baf86/logs/
+```
+
+The Model would report this URL in its status:
+
+```yaml
+kind: Model
+# ...
+status:
+  url: gs://abc123-substratus-artifacts/f94a0d128bcbd9c1b824e9e5572baf86
+```
+
+#### Reconcile Logic
+
+Pseudo-reconciler logic for a Model:
+
+```
+if .status.url != "" {
+  return
+}
+
+url := "{bucket}/{hash}"
+if sci.bucketObjectExists(url) {
```

Comment:
would this check if the directory in bucket is present and has files? or just if directory is present?

Reply:
Good question: also another consideration: the logic as written could result in an incomplete Model being loaded into the new cluster as a ready Model

Reply:
Updated to include a `completed.json` file check.